File: Javascript/ReverseLinkedListII.js
// Leetcode Problem - 92

var ListNode = function(val,next){
    this.val = val
    this.next = next
}

var reverseBetween = function(head,left,right){
    
    if (head === null || left === right){
        return head
    }

    const dummy = new ListNode(0,head)
    let prev = dummy

    for (let i=0;i<(left-1);i++){
        prev = prev.next
    }

    const curr = prev.next

    for (let i=0;i<(right-left);i++){
        const next_node = curr.next
        curr.next = next_node.next
        next_node.next = prev.next
        prev.next = next_node
    }

    return dummy.next
}

File: Python/ReverseLinkedListII.py
class ListNode:
    def __init__(self,val=0,next=None) -> None:
        self.val = val
        self.next = next

def reverseBetween(head,left,right):
    
    if head == None or left == right:
        return head
    
    dummy = ListNode(0,head)
    prev = dummy

    for i in range(left-1):
        prev = prev.next

    curr = prev.next

    for i in range(right-left):
        next_node = curr.next
        curr.next = next_node.next
        next_node.next = prev.next
        prev.next = next_node

    return dummy.next
